### ISSUE TITLE
Unfold doesn't pad the input and falils for small spans

### ISSUE THREAD

**Issue by vadimkantorov** [WiraDKP/pytorch_speaker_embedding_for_diarization#5]
https://github.com/WiraDKP/pytorch_speaker_embedding_for_diarization/blob/master/src/predict.py#L64 I think there should be an explicit F.pad call, at least when input is smaller than kernel_size

**Comment by vadimkantorov** [comment]
I added there:
```python
if mfcc.shape[-1] < self.max_frame:
    mfcc = torch.nn.functional.pad(mfcc, (0, self.max_frame - mfcc.shape[-1]))
```
prior to unfold

**Comment by WiraDKP** [comment]
Hi @vadimkantorov 
Thanks so much for your feedback.

The approach that I use in this code should not be applied on short audio. There is `min_dur` argument when sampling the data, and it should have removed them prior to training.

While I would also agree that some architecture design would downsample the mfcc too much that it would be shorter than the kernel itself. In my opinion, instead of padding, they should redesign the architecture instead because this approach (metric learning) would suffer if you add synthetic information such as zero padding.

Having that padding would make it work, codewise, but it would be confusing when the model does not work and the training is unstable.

I don't know. What do you think? @vadimkantorov

**Comment by vadimkantorov** [comment]
Even with this fix, I had it failing during some other padding operation: 200x200 reflection padding was applied to small input and failed

**Comment by vadimkantorov** [comment]
It's the code itself that does cutting, but doesn't follow the requirements of the downstream model.

My audios are a few minutes long.

**Comment by WiraDKP** [comment]
Hi @vadimkantorov 
Thanks for the follow up
The model itself is just a minimal example. You should design your own architecture as needed. What I have provided is just a 6 layers CNN.

Yes, the trim would follow the shortest mfcc in the batch. Sorry if it does not suit your needs. 

Can you please check if your shortest audio causes the problem?
Can you also try to increase the `min_dur` if the shortest audio do actually is the problem?
I would still not recommend you to pad especially when your audio is a minute long though.

I am refering to the `VCTKTripletDataloader`
```
min_frame = min([i.shape[-1] for i in X])
X = [i[:, :, :min_frame] for i in X]
```

Thank you.

**Comment by WiraDKP** [comment]
If your dataset consists of long audio, I would recommend you to modify the `VCTKTripletDataloader` so that it randomly crops the long mfcc with specified nuumber of frame. Hence, you'll have the benefit of long and variative audio. As currently in this code, it would trim the audio and does not utilize the benefit of your long audio.

Hope it helps.
Thanks
